src/textual/reactive.py

```diff
@@ -36,6 +43,43 @@ class TooManyComputesError(Exception):
     """Raised when an attribute has public and private compute methods."""
 
 
+async def await_watcher(obj: Reactable, awaitable: Awaitable[object]) -> None:
+    """Coroutine to await an awaitable returned from a watcher"""
+    _rich_traceback_omit = True
+    await awaitable
+    # Watcher may have changed the state, so run compute again
+    obj.post_message(events.Callback(callback=partial(Reactive._compute, obj)))
+
+
+def invoke_watcher(
+    watcher_object: Reactable,
+    watch_function: WatchCallbackType,
+    old_value: object,
+    value: object,
+) -> None:
+    """Invoke a watch function.
+
+    Args:
+        watcher_object: The object watching for the changes.
+        watch_function: A watch function, which may be sync or async.
+        old_value: The old value of the attribute.
+        value: The new value of the attribute.
+    """
+    _rich_traceback_omit = True
+    param_count = count_parameters(watch_function)
+    if param_count == 2:
+        watch_result = cast(WatchCallbackBothValuesType, watch_function)(
+            old_value, value
+        )
+    elif param_count == 1:
+        watch_result = cast(WatchCallbackNewValueType, watch_function)(value)
+    else:
+        watch_result = cast(WatchCallbackNoArgsType, watch_function)()
```

Comment:
Not sure I'm following this. The `param_count` can't be known until runtime as it gets computed dynamically, but the type checker runs before the program starts. So how can the conditionals + casts influence the type checkers behaviour if they run after the type checker?

Reply:
That's exactly how type narrowing works in general.
```py
x: int | None = ...
if x is not None:
    y = x + 3
```

The `if` only runs at runtime and yet the type checker understands that `if x is not None` is narrowing the type of `x`, effectively concluding that `x` must be an integer inside that `if`.

However, the type checker isn't smart enough to do type narrowing with the `if` with the `param_count` check.
So, it complains, even though we know the code runs precisely because at runtime we only run the branch that works.
Precisely because _these_ `if` do not influence the type checker, I add the casts to silence it.

Reply:
As you noted, the example you gave of `if x is not None` is not the same, since it's can be inferred statically.

I was mostly surprised about how it seems like a whole bunch of type definitions that are only used for casting, and their sole purpose seems to be to silence the type checker.

Not a blocker though, was just a curiosity :)

Reply:
Ah, yeah. I refactored the original `WatchCallbackType` because it was huge and it makes sense to split it into 3, depending on the number of arguments.
Then, I used that refactor to silence the type checker 🤷